sequencer/src/catchup.rs

```diff
@@ -0,0 +1,134 @@
+use crate::{
+    api::endpoints::{AccountQueryData, BlocksFrontier},
+    state::{BlockMerkleTree, FeeAccount, FeeMerkleCommitment},
+    Header,
+};
+use commit::Commitment;
+use futures::future::join_all;
+use hotshot_types::{data::ViewNumber, traits::node_implementation::ConsensusTime as _};
+use jf_primitives::merkle_tree::{ForgetableMerkleTreeScheme, MerkleTreeScheme as _};
+use serde::de::DeserializeOwned;
+use std::time::Duration;
+use surf_disco::Request;
+use tide_disco::error::ServerError;
+use url::Url;
+
+#[derive(Debug, Clone)]
+struct Client<ServerError> {
+    inner: surf_disco::Client<ServerError>,
+    url: Url,
+}
+
+impl Client<ServerError> {
+    pub fn new(url: Url) -> Self {
+        Self {
+            inner: surf_disco::Client::new(url.clone()),
+            url,
+        }
+    }
+
+    pub fn get<T: DeserializeOwned>(&self, route: &str) -> Request<T, ServerError> {
+        self.inner.get(route)
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct StatePeers {
+    clients: Vec<Client<ServerError>>,
+    interval: Duration,
+}
+
+impl StatePeers {
+    pub fn from_urls(urls: Vec<Url>) -> Self {
+        Self {
+            clients: urls.into_iter().map(Client::new).collect(),
+            interval: Duration::from_secs(1),
+        }
+    }
+
+    pub async fn fetch_accounts(
+        &self,
+        view: ViewNumber,
+        fee_merkle_tree_root: FeeMerkleCommitment,
+        accounts: impl IntoIterator<Item = FeeAccount>,
+    ) -> Vec<AccountQueryData> {
+        join_all(
+            accounts
+                .into_iter()
+                .map(|account| self.fetch_account(view, fee_merkle_tree_root, account)),
+        )
+        .await
+    }
+
+    async fn fetch_account(
+        &self,
+        view: ViewNumber,
+        fee_merkle_tree_root: FeeMerkleCommitment,
+        account: FeeAccount,
+    ) -> AccountQueryData {
+        if self.clients.is_empty() {
+            panic!("No peers to fetch account from");
```

Comment:
I think it'll be easier to debug config mistakes if we panic in the constructor rather than here in what is probably a background task

Reply:
Yes I will have a think about how to improve this. I think it's good to have way to run without peers. Maybe just making the state peers an `Option` is the simplest way to go.

Reply:
The thing is if you run without peers and your node falls behind you're unable to vote ever again. I think we will just have to require at least one peer in order to run the node. And we can give our partners the URLs of our nodes to use for catchup.

I think ideally every node should have the URL of `f+1` other nodes, which guarantees (I think) that at least one of your peers will have the state for every view

Reply:
Yeah I agree it may be a good idea to enforce this on startup of a production node but we probably don't want to enforce it for every test deployment.

I think it would be good if

1. We can enforce running with likely correctly configured state peers.
1. We can run without real state peers for tests.
1. If we run tests we can either make it impossible accidentally to look up state from peers or panic if we do so. This avoids having many tests hang forever. Already happened to me :laughing: 

The simplest thing I can think of that achieves this is to

1. Enforce state peers aren't empty in regular constructor (as suggested).
2. Hide a test constructor behind `testing` feature flag to create empty state peer.
3. panic if we end up doing lookups with empty state peers

I do think this is kind of ugly though. Let me know if you have better idea.

If we go down the route of making a trait and a test + real implementation I think it's quite a lot of change required just to avoid a few panics in the code, so not sure it's worth it.

What I think I want to do is having a test implementation of `StatePeers` that doesn't use the network but takes as constructor input the merkle trees and then on lookup looks up value in it's own data structure and use that for tests. But again it seems a bit involved and not sure if it's worth doing right now.

Reply:
I am going to need something along those lines to get the restart test working, basically to represent an archival node without actually having state storage. I do think having a `StateCatchup` will be useful for these sorts of things. Is it maybe not so much of a change if we `Box<dyn StateCatchup>` instead of trying to make `InstanceState` generic?